### DISCUSSION TITLE
Performance issue with DrawImage() in game.Draw()

### DISCUSSION THREAD

**Discussion by TLINDEN** [hajimehoshi/ebiten#3002] [Q&A]
### Ebitengine Version

2.7.4

### Operating System

- [ ] Windows
- [ ] macOS
- [X] Linux
- [ ] FreeBSD
- [ ] OpenBSD
- [ ] Android
- [ ] iOS
- [ ] Nintendo Switch
- [ ] PlayStation 5
- [ ] Xbox
- [ ] Web Browsers

### Go Version (`go version`)

1.22.1

### What steps will reproduce the problem?

I wrote a small game of life clone using Ebitengine. It works nicely so far. But when I increase the number of cells above some certain amount, then CPU performance detoriates dramatically. FPS goes down to <5. If I enable CPU profiling on the game, the whole system freezes and needs a hard reset.

### What is the expected result?

No issue with high cell (that is: image) counts.

### What happens instead?

FPS goes down rapidly and the game becomes unresponsive.

### Anything else you feel useful to add?

The source of the game in question is here: [golsky](https://github.com/TLINDEN/golsky/). I was unable to actually get a non-zero profiling file using the ebitengine main loop. Therefore I created a [fake main loop](https://github.com/TLINDEN/golsky/blob/main/main.go#L56). Profiling can be initiated with these options:
```
      --profile-draw               profile draw method (default false)
      --profile-file string        enable profiling
      --profile-max-loops int      how many loops to execute (default 10) (default 10)
```

This is an example call which causes the FPS problem:
```
./golsky -W800 -H800 --profile-file cpu.profile --profile-max-loops 50 --profile-draw
```

My first thought was that my own code is the culprit, of course. But when I looked into the profiling results, it looked different:

![profile-top](https://github.com/hajimehoshi/ebiten/assets/1322812/b6363044-51c1-406a-af87-a65e8d23d3c4)

![profile-graph](https://github.com/hajimehoshi/ebiten/assets/1322812/897f7b6f-215f-416f-bca2-a61c591a1fe8)

So, in the game I [pre-compute](https://github.com/TLINDEN/golsky/blob/main/scene-play.go#L545) the tile images and in `game.Draw()` I just [put](https://github.com/TLINDEN/golsky/blob/main/scene-play.go#L427)  these images onto the screen image.

Ok, I'm still new to game development (and even golang for that matter), so I'm pretty sure there might be lots of things I could improve in my code. But `pprof` says something different.

**Comment by hajimehoshi** [comment]
(I've moved this to the discussion forum. If I found an actionable item, I'll create a new issue)

Apparently atlas.DrawTriangles creates many objects, but I was wondering if there are other objects in your program.

**Reply by TLINDEN** [reply]
Ok, I tried to make a somewhat minimal version. This one just puts random slots onto the grid:

https://gist.github.com/TLINDEN/fae8f9d3666c01d074c394330d1e1d42

Flame graph:
![minimalprof](https://github.com/hajimehoshi/ebiten/assets/1322812/36e4b07e-3446-4fb7-ac32-c5692e62b367)

**Reply by hajimehoshi** [reply]
Thank you for minimizing the case, but unfortunately this doesn't reproduce the same issue: DrawTriangle is no longer an allocation-heavy part and the situation is different...

**Reply by hajimehoshi** [reply]
> Apparently atlas.DrawTriangles creates many objects, but I was wondering if there are other objects in your program.

So, I suspect that there were memory consumers in your original program other than Ebitengine. 🤔

**Reply by TLINDEN** [reply]
I don't know. If I modify `game.Draw()` so that it just returns instead of drawing anything, it looks like this:

![nodraw](https://github.com/hajimehoshi/ebiten/assets/1322812/1031baba-1091-41e6-95b4-7e2e32a61b4e)

And in this situation I get roughly 50 FPS. So, to me it looks like CPU is "eaten" inside `Draw`.

Also, memory consumption is pretty low and constant:
```
top:
    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND                                 
  74075 scip      20   0 5548112   1.2g  60928 S 106.0   3.7   1:42.98 golsky 
```
Not a memory problem.

Besides - I also changed the code to use a Shader to draw the tiles instead of images. But the response is the very same, just this time `DrawRectShader()` eats most of the CPU time:

![shader-cpu](https://github.com/hajimehoshi/ebiten/assets/1322812/55d766d2-2714-42d0-83b7-6ffe1fcbe362)

**Reply by hajimehoshi** [reply]
> Not a memory problem.

From the first graph, I thought the original issue was that memory allocations happened too often and GC took long time. Actually `runtime.scanobject` is the heaviest task in the first attempt, right?

**Reply by hajimehoshi** [reply]
I looked at your minimized test case (https://gist.github.com/TLINDEN/fae8f9d3666c01d074c394330d1e1d42), and apparently this calls DrawImage 800x600 = 480000 times, and yes, this should be very heavy. There might be things Ebitengine can do, but in the first place, I'd recommend reducing DrawImage calls in your games.

**Reply by hajimehoshi** [reply]
In general, you can reduce DrawImage calls by e.g. not calling it when the rendering object is out of the screen. I've not looked into your original game carefully yet, but would it be possible?

**Reply by TLINDEN** [reply]
Well, I am indeed using a camera view to support zooming. I haven't just not yet figured out, how to start in a zoomed state. However, even if I did that, the user might still want to zoom out up to this slow point.

What do you think, would it be faster to use `vector.DrawFilledRect()`?

Or, is it possible to supply a slice as a uniform to a shader?

**Reply by TLINDEN** [reply]
I followed your advice and [added some caching](https://github.com/TLINDEN/golsky/commit/8cb5585456222d15b74d274f3c9bde3804771c0e): on game startup I paint the whole grid onto a cache image including all dead cells. Then in `Draw()` I only paint life cells (and evolution traces ir requested) but not dead cells anymore. Since the majority of cells are dead anyway, this reduces the calls to `DrawImage()` dramatically. With this measure the game runs fluid at 800x600 cells. Even 1200x800 is still working ok, but above it's getting slow. But for my purposes this is good enough for the time being.

Maybe drawing directly may be even faster, I'll try this out tomorrow.

**Reply by hajimehoshi** [reply]
> However, even if I did that, the user might still want to zoom out up to this slow point.

Good question. When zooming out, do you really need rendering all the small things? If the objects are static, I'd use an offscreen rendering (rendering tiles onto an offscreen and use it).

> What do you think, would it be faster to use vector.DrawFilledRect()?

As long as the number of vertices doesn't change, the performance is the same.

> Or, is it possible to supply a slice as a uniform to a shader?

It is possible to supply a slice as a uniform variable. (`[]float32 or []int32`).

> I followed your advice and added some caching ...

This is what I called "an offscreen", and I think this seems the way to go. 👍

**Reply by TLINDEN** [reply]
>>What do you think, would it be faster to use vector.DrawFilledRect()?
>As long as the number of vertices doesn't change, the performance is the same.

I have tested it, the variant using `DrawImage()` is faster, but not by much.

> It is possible to supply a slice as a uniform variable. ([]float32 or []int32).

But only one dimension, but not `[][]int32`? Because if it would be possible, I could just use one shader for the whole grid, which gets the current grid setup as a uniform and updates all cells accordingly - et voilà :)

While I'm thinking about it, what I could DO with one dimesion is to use one shader per row. Would still be better than lots of single DrawImage calls...

> This is what I called "an offscreen", and I think this seems the way to go.

Yes, but I only do it for a part of the screen (background/grid + dead cells), then I copy the offscreen image into the world image and finally draw the life cells to it.

Would it be faster to do the last step offscreen as well and copying it only at the end?

**Reply by hajimehoshi** [reply]
> But only one dimension, but not [][]int32? Because if it would be possible, I could just use one shader for the whole grid, which gets the current grid setup as a uniform and updates all cells accordingly - et voilà :)

Right, two dimensional slices are not supported so far.

> While I'm thinking about it, what I could DO with one dimesion is to use one shader per row. Would still be better than lots of single DrawImage calls...

If the purpose is to reduce the number of `DrawImage` calls, I'd use `DrawTriangles` or `DrawTrianglesShader` by creating your own vertices. This means you will do a very similar thing to what Ebitengine does for `DrawImage` internally, but you can get your own optimized logic. What do you think?

> Would it be faster to do the last step offscreen as well and copying it only at the end?

I'm not sure I understand what you did. What kind of items and images are there in your game?

* background
* grid
* dead cells
* live cells

? I though you rendered everything but live cells are rendered onto an offscreen once, and then rendered a part of the offscreen onto the final screen, and then rendered the live cells onto the final screen.

**Reply by hajimehoshi** [reply]
The indices would be like 0, 1, 2, 1, 2, 3, 4, 5, 6, 5, 6, 7, ... for rectangles. 6 indices should be added for 4 vertices = 2 triangles = 1 rectangle. In your code, 3 indices are added for 1 vector, and this is too much.

**Reply by TLINDEN** [reply]
Yes, I think I fixed the triangles generation. But as already discussed on Discord, at a certain amount of cells the output gets [more and more artifacts](https://discord.com/channels/842049801528016967/842056438632939570/1245345744135786559). [Source for the DrawTriangles variant](https://github.com/TLINDEN/testgol/tree/main).

So, as suggested I also build an example which uses `WritePixels()`: https://github.com/TLINDEN/testgol/tree/wrpixels. This works fine and is even faster (and without artifacts). 

But even by using WritePixels I'm reaching a performance barrier > 1000x1000 cells. According to pprof measurements, this time it's not related to any ebitengine function calls, but seems to be caused solely by the functin which calculates the vertices for WritePixels: [`UpdatePixels()`](https://github.com/TLINDEN/testgol/blob/wrpixels/main.go#L198).

So, I think at this point you're out of the issue, but I'd like to ask anyway: are there any further enhancements you might be able to suggest? I'm asking because I have seen much larger game of life grids in the past (and I'm talking about the 2 decades ago past).

However, thank you very much for your time and patiene! I know that this is not a given.

best,
Tom

**Reply by hajimehoshi** [reply]
Thanks!

> So, I think at this point you're out of the issue, but I'd like to ask anyway: are there any further enhancements you might be able to suggest? I'm asking because I have seen much larger game of life grids in the past (and I'm talking about the 2 decades ago past).

TBH, I don't have a good idea. For a life game, you can utilize GPU to calculate the next state by utilizing parallelization.  Go (and Ebitengine) doesn't have a good API or a library to do GPU calculation. Maybe Kage might be available but I have not tried for this purpose :-)

**Comment by TLINDEN** [comment]
Basically I draw a background and then just colored tiles with a little distance between them to simulate grid lines (if enabled). That's really all to it.

**Reply by hajimehoshi** [reply]
Is this the reply to https://github.com/hajimehoshi/ebiten/discussions/3002#discussioncomment-9569772?

**Reply by hajimehoshi** [reply]
I want a more minimized test case to reproduce this memory allocation. With some examples like sprites, I failed to reproduce this.
